src/UtilitiesService.php

```diff
@@ -0,0 +1,244 @@
+        '#type' => 'checkbox',
+        '#title' => $this->t('Use Canada Post sitewide API settings'),
+        '#description' => $this->t('The Canada Post @url will be used when fetching rates and tracking details.
+          <br \><strong>Uncheck</strong> this box if you\'d like to use a different account when fetching rates and tracking details for orders from this store.', [
+            '@url' => Link::fromTextAndUrl(
+              $this->t('sitewide API settings'),
+              Url::fromRoute('commerce_canadapost.settings_form')
+            )->toString(),
+          ]
+        ),
+        '#default_value' => $store_settings_set,
+      ];
+    }
+
+    // Fetch the Canada Post API settings.
+    $api_settings = $this->getApiSettings($store);
+
+    $form['api']['customer_number'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Customer number'),
+      '#default_value' => $api_settings['customer_number'],
+      '#required' => TRUE,
+    ];
+    $form['api']['username'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Username'),
+      '#default_value' => $api_settings['username'],
+      '#required' => TRUE,
+    ];
+    $form['api']['password'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Password'),
+      '#default_value' => $api_settings['password'],
+      '#required' => TRUE,
+    ];
+    $form['api']['contract_id'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Contract ID'),
+      '#default_value' => $api_settings['contract_id'],
+    ];
+    $form['api']['mode'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Mode'),
+      '#default_value' => $api_settings['mode'],
+      '#options' => [
+        'test' => $this->t('Test'),
+        'live' => $this->t('Live'),
+      ],
+      '#required' => TRUE,
+    ];
+    $form['api']['log'] = [
+      '#type' => 'checkboxes',
+      '#title' => $this->t('Log the following messages for debugging'),
+      '#options' => [
+        'request' => $this->t('API request messages'),
+        'response' => $this->t('API response messages'),
+      ],
+      '#default_value' => $api_settings['log'],
+    ];
+
+    // Alter the fields if we're in the store form.
+    if ($store) {
+      $this->alterApiFormFields($form);
+    }
+
+    return $form;
+  }
+
+  /**
+   * Alter the Canada Post API settings form fields if we're in the store form.
+   *
+   * @param array $form
+   *   The form array.
+   */
+  protected function alterApiFormFields(array &$form) {
+    // Fields should be visible only if the use_site_settings checkbox is
+    // unchecked.
+    $states = [
+      'visible' => [
+        ':input[name="use_sitewide_settings"]' => [
+          'checked' => FALSE,
+        ],
+      ],
+      'required' => [
+        ':input[name="use_sitewide_settings"]' => [
+          'checked' => FALSE,
+        ],
+      ],
+    ];
+    foreach ($this->getApiKeys() as $key) {
```

Comment:
there is probably a fancy `array_walk` function ( i still find very tough to follow) that dimitris will recommend to shorten this up

Reply:
I hate array_walks for the same reason, I could never follow it properly. But, I will do it for Dimitris!